src/colorizer/VectorField.ts

```diff
@@ -0,0 +1,313 @@
+   * Relative frame coordinates are in a range from `[-0.5, 0.5]`, and
+   * screen space pixel coordinates match the onscreen canvas pixels.
+   */
+  private relativeFrameCoordsToScreenSpacePx(vector: ArrayVector3): ArrayVector3 {
+    return [
+      vector[0] * this.frameToCanvasCoordinates.x * this.canvasResolution.x,
+      vector[1] * this.frameToCanvasCoordinates.y * this.canvasResolution.y,
+      vector[2],
+    ];
+  }
+
+  private getArrowHeadVertices(normVectorStart: ArrayVector3, normVectorEnd: ArrayVector3): Float32Array {
+    //    o <- normVectorEnd
+    //   /|\
+    //  / | \  <- arrow head maintains onscreen pixel length with zoom
+    //    |
+    //    |
+    //    o <- normVectorStart
+
+    // Draw the arrow heads. These are two line segments for each, so there's a total of four additional vertices we need to add.
+    const vertices = new Float32Array((VERTICES_PER_VECTOR_LINE - 2) * VERTEX_LENGTH);
+
+    // TODO: Handle scaling and arrow head length in a vertex shader.
+    // There's extra work being done here to keep the arrow head length constant with zoom, but it also
+    // means that currently this is recalculated EVERY time the zoom changes.
+    const screenSpaceStartPx = this.relativeFrameCoordsToScreenSpacePx(normVectorStart);
+    const screenSpaceEndPx = this.relativeFrameCoordsToScreenSpacePx(normVectorEnd);
+    const screenSpaceDeltaPx = [
+      screenSpaceEndPx[0] - screenSpaceStartPx[0],
+      screenSpaceEndPx[1] - screenSpaceStartPx[1],
+    ];
+    const vectorLengthPx = Math.sqrt(
+      screenSpaceDeltaPx[0] * screenSpaceDeltaPx[0] + screenSpaceDeltaPx[1] * screenSpaceDeltaPx[1]
+    );
+
+    const vectorAngle = Math.atan2(screenSpaceDeltaPx[1], screenSpaceDeltaPx[0]) + Math.PI;
+    const arrowAngle1 = vectorAngle + arrowStyle.headAngleRads;
+    const arrowAngle2 = vectorAngle - arrowStyle.headAngleRads;
+    const arrowHead1: ArrayVector3 = [Math.cos(arrowAngle1), Math.sin(arrowAngle1), 0];
+    const arrowHead2: ArrayVector3 = [Math.cos(arrowAngle2), Math.sin(arrowAngle2), 0];
+
+    const lengthPx = Math.min(arrowStyle.maxHeadLengthPx * 2, vectorLengthPx);
+
+    // Keep arrow head length constant relative to onscreen pixels
+    // Arrow heads are in screen space pixel coordinates, so we divide by canvas resolution to get relative canvas coordinates,
+    // and then divide again by `frameToCanvasCoordinates` to get relative frame coordinates.
+    arrowHead1[0] =
+      (arrowHead1[0] * lengthPx) / this.frameToCanvasCoordinates.x / this.canvasResolution.x + normVectorEnd[0];
+    arrowHead1[1] =
+      (arrowHead1[1] * lengthPx) / this.frameToCanvasCoordinates.y / this.canvasResolution.y + normVectorEnd[1];
+    arrowHead2[0] =
+      (arrowHead2[0] * lengthPx) / this.frameToCanvasCoordinates.x / this.canvasResolution.x + normVectorEnd[0];
+    arrowHead2[1] =
+      (arrowHead2[1] * lengthPx) / this.frameToCanvasCoordinates.y / this.canvasResolution.y + normVectorEnd[1];
```

Comment:
I would factor this out. First, I would have a utility function that just scales a vector to a given length and returns the new vector. second, I might also have a "scaleArrowHead" function that pulls this code out of this larger function